File: flutter/sherpa_onnx/lib/src/spoken_language_identification.dart
// Copyright (c)  2024  Xiaomi Corporation
import 'dart:ffi';

import 'package:ffi/ffi.dart';

import './offline_stream.dart';
import './sherpa_onnx_bindings.dart';
import './utils.dart';
import './spoken_language_identification_config.dart';

export './spoken_language_identification_config.dart';

/// Spoken language identifier.
class SpokenLanguageIdentification {
  SpokenLanguageIdentification.fromPtr(
      {required this.ptr, required this.config});

  SpokenLanguageIdentification._({required this.ptr, required this.config});

  /// Release the native language identifier.
  void free() {
    if (SherpaOnnxBindings.sherpaOnnxDestroySpokenLanguageIdentification ==
        null) {
      throw Exception("Please initialize sherpa-onnx first");
    }

    if (ptr == nullptr) {
      return;
    }
    SherpaOnnxBindings.sherpaOnnxDestroySpokenLanguageIdentification?.call(ptr);
    ptr = nullptr;
  }

  /// Create a language identifier from [config].
  factory SpokenLanguageIdentification(
      SpokenLanguageIdentificationConfig config) {
    final c = convertConfig(config);

    if (SherpaOnnxBindings.sherpaOnnxCreateSpokenLanguageIdentification ==
        null) {
      freeConfig(c);
      throw Exception("Please initialize sherpa-onnx first");
    }

    final ptr = SherpaOnnxBindings.sherpaOnnxCreateSpokenLanguageIdentification
            ?.call(c) ??
        nullptr;

    if (ptr == nullptr) {
      freeConfig(c);
      throw Exception(
          "Failed to create spoken language identification. Please check your config");
    }

    freeConfig(c);

    return SpokenLanguageIdentification._(ptr: ptr, config: config);
  }

  static Pointer<SherpaOnnxSpokenLanguageIdentificationConfig> convertConfig(
      SpokenLanguageIdentificationConfig config) {
    final c = calloc<SherpaOnnxSpokenLanguageIdentificationConfig>();

    c.ref.whisper.encoder = config.whisper.encoder.toNativeUtf8();
    c.ref.whisper.decoder = config.whisper.decoder.toNativeUtf8();
    c.ref.whisper.tailPaddings = config.whisper.tailPaddings;

    c.ref.numThreads = config.numThreads;
    c.ref.debug = config.debug ? 1 : 0;
    c.ref.provider = config.provider.toNativeUtf8();

    return c;
  }

  static void freeConfig(
      Pointer<SherpaOnnxSpokenLanguageIdentificationConfig> c) {
    malloc.free(c.ref.whisper.encoder);
    malloc.free(c.ref.whisper.decoder);
    malloc.free(c.ref.provider);
    malloc.free(c);
  }

  /// Create an offline stream for one audio clip.
  OfflineStream createStream() {
    if (SherpaOnnxBindings
            .sherpaOnnxSpokenLanguageIdentificationCreateOfflineStream ==
        null) {
      throw Exception("Please initialize sherpa-onnx first");
    }

    if (ptr == nullptr) {
      throw Exception("Failed to create offline stream");
    }

    final p = SherpaOnnxBindings
            .sherpaOnnxSpokenLanguageIdentificationCreateOfflineStream
            ?.call(ptr) ??
        nullptr;

    if (p == nullptr) {
      throw Exception("Failed to create offline stream");
    }

    return OfflineStream(ptr: p);
  }

  /// Compute the spoken language for [stream].
  SpokenLanguageIdentificationResult compute(OfflineStream stream) {
    if (SherpaOnnxBindings.sherpaOnnxSpokenLanguageIdentificationCompute ==
        null) {
      throw Exception("Please initialize sherpa-onnx first");
    }

    if (ptr == nullptr || stream.ptr == nullptr) {
      return const SpokenLanguageIdentificationResult(lang: '');
    }

    final result = SherpaOnnxBindings
            .sherpaOnnxSpokenLanguageIdentificationCompute
            ?.call(ptr, stream.ptr) ??
        nullptr;

    if (result == nullptr) {
      return const SpokenLanguageIdentificationResult(lang: '');
    }

    final lang = toDartString(result.ref.lang);

    SherpaOnnxBindings.sherpaOnnxDestroySpokenLanguageIdentificationResult
        ?.call(result);

    return SpokenLanguageIdentificationResult(lang: lang);
  }

  Pointer<SherpaOnnxSpokenLanguageIdentification> ptr;
  SpokenLanguageIdentificationConfig config;
}

File: sherpa-onnx/rust/sherpa-onnx/src/spoken_language_identification.rs
//! Spoken language identification.
//!
//! This module identifies the language spoken in an audio clip using the
//! Whisper-based language ID API. See
//! [`rust-api-examples/examples/spoken_language_identification.rs`](https://github.com/k2-fsa/sherpa-onnx/blob/master/rust-api-examples/examples/spoken_language_identification.rs)
//! for a complete example.
//!
//! # Example
//!
//! ```no_run
//! use sherpa_onnx::{
//!     SpokenLanguageIdentification, SpokenLanguageIdentificationConfig,
//!     SpokenLanguageIdentificationWhisperConfig, Wave,
//! };
//!
//! let wave = Wave::read("./test.wav").expect("read wave");
//! let config = SpokenLanguageIdentificationConfig {
//!     whisper: SpokenLanguageIdentificationWhisperConfig {
//!         encoder: Some("./sherpa-onnx-whisper-tiny/encoder.int8.onnx".into()),
//!         decoder: Some("./sherpa-onnx-whisper-tiny/decoder.int8.onnx".into()),
//!         tail_paddings: 0,
//!     },
//!     ..Default::default()
//! };
//!
//! let slid = SpokenLanguageIdentification::create(&config).expect("create");
//! let stream = slid.create_stream();
//! stream.accept_waveform(wave.sample_rate(), wave.samples());
//! let result = slid.compute(&stream).expect("compute");
//! println!("{}", result.lang);
//! ```

use crate::offline_asr::OfflineStream;
use crate::utils::to_c_ptr;
use sherpa_onnx_sys as sys;
use std::ffi::{CStr, CString};

#[derive(Clone, Debug, Default)]
/// Whisper model configuration for spoken language identification.
pub struct SpokenLanguageIdentificationWhisperConfig {
    pub encoder: Option<String>,
    pub decoder: Option<String>,
    pub tail_paddings: i32,
}

impl SpokenLanguageIdentificationWhisperConfig {
    fn to_sys(
        &self,
        cstrings: &mut Vec<CString>,
    ) -> sys::SpokenLanguageIdentificationWhisperConfig {
        sys::SpokenLanguageIdentificationWhisperConfig {
            encoder: to_c_ptr(&self.encoder, cstrings),
            decoder: to_c_ptr(&self.decoder, cstrings),
            tail_paddings: self.tail_paddings,
        }
    }
}

#[derive(Clone, Debug)]
/// Top-level configuration for [`SpokenLanguageIdentification`].
pub struct SpokenLanguageIdentificationConfig {
    pub whisper: SpokenLanguageIdentificationWhisperConfig,
    pub num_threads: i32,
    pub debug: bool,
    pub provider: Option<String>,
}

impl Default for SpokenLanguageIdentificationConfig {
    fn default() -> Self {
        Self {
            whisper: Default::default(),
            num_threads: 1,
            debug: false,
            provider: Some("cpu".to_string()),
        }
    }
}

impl SpokenLanguageIdentificationConfig {
    fn to_sys(&self, cstrings: &mut Vec<CString>) -> sys::SpokenLanguageIdentificationConfig {
        sys::SpokenLanguageIdentificationConfig {
            whisper: self
                .whisper
                .to_sys(cstrings),
            num_threads: self.num_threads,
            debug: self.debug as i32,
            provider: to_c_ptr(&self.provider, cstrings),
        }
    }
}

#[derive(Clone, Debug)]
/// Result returned by [`SpokenLanguageIdentification::compute`].
pub struct SpokenLanguageIdentificationResult {
    pub lang: String,
}

/// Spoken language identifier.
pub struct SpokenLanguageIdentification {
    ptr: *const sys::SpokenLanguageIdentification,
}

unsafe impl Send for SpokenLanguageIdentification {}

impl SpokenLanguageIdentification {
    /// Create a language identifier from [`SpokenLanguageIdentificationConfig`].
    pub fn create(config: &SpokenLanguageIdentificationConfig) -> Option<Self> {
        let mut cstrings = Vec::new();
        let sys_config = config.to_sys(&mut cstrings);
        let ptr = unsafe { sys::SherpaOnnxCreateSpokenLanguageIdentification(&sys_config) };
        if ptr.is_null() {
            None
        } else {
            Some(Self { ptr })
        }
    }

    /// Create an offline stream for one audio clip.
    pub fn create_stream(&self) -> OfflineStream {
        let ptr =
            unsafe { sys::SherpaOnnxSpokenLanguageIdentificationCreateOfflineStream(self.ptr) };
        OfflineStream { ptr }
    }

    /// Compute the spoken language for `stream`.
    pub fn compute(&self, stream: &OfflineStream) -> Option<SpokenLanguageIdentificationResult> {
        unsafe {
            let p = sys::SherpaOnnxSpokenLanguageIdentificationCompute(self.ptr, stream.ptr);
            if p.is_null() {
                return None;
            }

            let ans = SpokenLanguageIdentificationResult {
                lang: if (*p)
                    .lang
                    .is_null()
                {
                    String::new()
                } else {
                    CStr::from_ptr((*p).lang)
                        .to_string_lossy()
                        .into_owned()
                },
            };

            sys::SherpaOnnxDestroySpokenLanguageIdentificationResult(p);
            Some(ans)
        }
    }
}

impl Drop for SpokenLanguageIdentification {
    fn drop(&mut self) {
        unsafe {
            if !self
                .ptr
                .is_null()
            {
                sys::SherpaOnnxDestroySpokenLanguageIdentification(self.ptr);
            }
        }
    }
}
